lib/std/interface.zig

```diff
@@ -0,0 +1,510 @@
+
+        if (type_info.Fn.args.len == 0) {
+            @compileError("Virtual function '" ++ field.name ++ "' must have at least one argument.");
+        }
+
+        const arg_type = type_info.Fn.args[0].arg_type.?;
+        if (arg_type != *SelfType and arg_type != *const SelfType) {
+            @compileError("Virtual function's '" ++ field.name ++ "' first argument must be *SelfType or *const SelfType");
+        }
+    }
+}
+
+fn vtableHasMethod(comptime VTableT: type, comptime name: []const u8, is_optional: *bool) bool {
+    for (std.meta.fields(VTableT)) |field| {
+        if (std.mem.eql(u8, name, field.name)) {
+            is_optional.* = trait.is(.Optional)(field.field_type);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+fn VTableReturnType(comptime VTableT: type, comptime name: []const u8) type {
+    for (std.meta.fields(VTableT)) |field| {
+        if (std.mem.eql(u8, name, field.name)) {
+            const is_optional = trait.is(.Optional)(field.field_type);
+            // TODO: Do I need to do smth different for async?
+            if (is_optional) {
+                return ?@typeInfo(std.meta.Child(field.field_type)).Fn.return_type.?;
+            }
+
+            return @typeInfo(field.field_type).Fn.return_type orelse noreturn;
+        }
+    }
+
+    @compileError("VTable type '" ++ @typeName(VTableT) ++ "' has no virtual function '" ++ name ++ "'.");
+}
+
+pub fn Interface(comptime VTableT: type, comptime StorageT: type) type {
+    comptime checkVtableType(VTableT);
+
+    return struct {
+        vtable_ptr: *const VTableT,
+        storage: StorageT,
+
+        const Self = @This();
+
+        pub fn init(args: var) !Self {
+            const ImplType = PtrChildOrSelf(@TypeOf(args.@"0"));
+
+            return Self{
+                .vtable_ptr = &comptime makeVTable(VTableT, ImplType),
+                .storage = try StorageT.init(args),
+            };
+        }
+
+        pub fn initWithVTable(vtable_ptr: *const VTableT, args: var) !Self {
+            return .{
+                .vtable_ptr = vtable_ptr,
+                .storage = try StorageT.init(args),
+            };
+        }
+
+        pub fn call(self: Self, comptime name: []const u8, args: var) VTableReturnType(VTableT, name) {
+            comptime var is_optional = true;
+            comptime assert(vtableHasMethod(VTableT, name, &is_optional));
+
+            const fn_ptr = if (is_optional) blk: {
+                const val = @field(self.vtable_ptr, name);
+                if (val) |v| break :blk v;
+                return null;
+            } else @field(self.vtable_ptr, name);
+
+            const self_ptr = self.storage.getSelfPtr();
+            const new_args = .{self_ptr};
+
+            return @call(.{}, fn_ptr, new_args ++ args);
+        }
+
+        pub fn deinit(self: Self) void {
+            self.storage.deinit();
+        }
+    };
+}
+
+var testing_allocator = &std.testing.allocator_instance.allocator;
```

Comment:
Shouldn't this just be `testing.allocator` inside of your test?

Reply:
Yes, missed that declaration somehow >.<

Reply:
@daurnimator Using std.testing.allocator directly  on line 499 triggers #4597, will be documented in comments